relations/src/r1cs/constraint_system.rs

```diff
@@ -309,91 +310,111 @@ impl<F: Field> ConstraintSystem<F> {
             return;
         }
 
-        let used_times = self.lc_num_times_used(true);
-
-        // step 2: Modify the lcs accordingly
-        let mut outlined_lcs = BTreeMap::new();
-
-        let mut additional_a_constraints = Vec::<LcIndex>::new();
-        let mut additional_c_constraints_witness_assignments = Vec::<usize>::new();
+        let mut num_times_used = self.lc_num_times_used(false);
 
+        let mut new_witness_linear_combinations = Vec::new();
+        let mut new_witness_indices = Vec::new();
+
+        // `new_lc_map` stores the modified linear combinations.
+        //
+        // This loop goes through all the LCs in the map, starting from
+        // the early ones, and decides whether or not to dedicate a witness
+        // variable for this LC.
+        //
+        // If true, the LC is replaced with 1 * this witness variable.
+        // Otherwise, the LC is inlined.
+        //
+        let mut new_lc_map = BTreeMap::new();
         for (&index, lc) in self.lc_map.iter() {
-            let mut outlined_lc = LinearCombination::new();
+            // Inline the LC.
+            let mut inlined_lc = LinearCombination::new();
             for &(coeff, var) in lc.iter() {
                 if var.is_lc() {
                     let lc_index = var.get_lc_index().expect("should be lc");
+
                     // If `var` is a `SymbolicLc`, fetch the corresponding
                     // inlined LC, and substitute it in.
-                    let lc = outlined_lcs.get(&lc_index).expect("should be inlined");
-                    outlined_lc.extend((lc * coeff).0.into_iter());
+                    //
+                    // We have the guarantee that `lc_index` must exist in
+                    // `new_lc_map` since a LC can only depend on other
+                    // LCs with lower indices, which we have modified.
+                    //
+                    let lc = new_lc_map.get(&lc_index).expect("should be inlined");
+                    inlined_lc.extend((lc * coeff).0.into_iter());
+
+                    num_times_used[lc_index.0] -= 1;
+                    if num_times_used[lc_index.0] == 0 {
+                        // This lc is not used any more, so remove it.
+                        new_lc_map.remove(&lc_index);
```

Comment:
Hmm doesn't this delete LCs when they're made into a new variable? For inlining it doesn't matter, because they're substituted directly into the actual constraints. But here, if the LC is "outlined", we cannot delete it from `new_lc_map`, no?

Reply:
Actually, we can. 

If an LC is "outlined", then its entry in `new_lc_map` is now `(1, the new witness variable)`. 

When it is not used anymore (aka, triggering this code), it means that all the subsequent LCs that use this LC has inlined this LC, by extending a term `(1 * coeff, the new witness variable)` into themselves. So, it can be safely deleted from the `new_lc_map`.

And this does not affect the code at about line 411, which creates additional constraints. This is because we copy the "inlined" LC data (not its index) in `new_witness_linear_combinations`.

Reply:
Ah I misunderstood; this works because the witness *variable* exists, and a corresponding constraint, and so the LC is useless. cool!

In that case, we should definitely extract this common logic between `inline_all_lcs` and `outline_lcs` (by taking in `new_lc_maps`), and leave the above explanation as a comment above the use in `outline_lcs`.

Reply:
Like this? Or more than that?
```
// All the LCs depending on `lc_index` have inlined it.
//
// If `lc_index` is outlined, then its witness variable has 
// replaced the original LC.
//
// This lc is not used any more, so remove it.
```

Reply:
How about something like this?

```rust 
// Deletion is safe for both outlining and inlining:
// * Inlining: the LC is substituted directly into all use 
//   sites, and so once it is fully inlined, it is redundant.
// * Outlining: the LC is associated with a new variable `w`,
//   and a new constraint of the form `lc_data * 1 = w`, where
//   `lc_data` is the actual data in the linear combination.
//   Furthermore, we replace its entry in `new_lc_map` with `(1, w)`.
//   Once `w` is fully inlined, then we can delete the entry from `new_lc_map`
```

Reply:
Needs a period at the end of the comments. 

Add this comment to BOTH `inline_all_lcs` and `outline_lcs`, or?

Reply:
And it makes sense!

Reply:
Let's extract the common logic into a method, and then call this method from `inline_all_lcs` and `outline_lcs`?

Reply:
Let me have a try on the common logic. Very likely, `inline_all_lcs` and `outline_lcs` is providing a stateful function to this common interface. The common interface asks this function to rewrite each LC, and this function can keep a state.

Reply:
Yup!

Reply:
This doesn't matter for soundness right? Its just for efficiency?

Reply:
Yes, for memory consumption.